Baikal/PostEffects/ML/denoiser.cpp

```diff
@@ -22,5 +52,150 @@ namespace Baikal
+
+            m_host_cache = std::move(std::make_unique<std::uint8_t[]>(elems_count));
+
+            // compute memory layout
+            switch (inputs)
+            {
+                case MLDenoiserInputs::kColorDepthNormalGloss7:
+                {
+                    m_layout.emplace(OutputType::kColor, 0u);
+                    m_layout.emplace(OutputType::kDepth, 3 * width * height);
+                    m_layout.emplace(OutputType::kViewShadingNormal, 4 * width * height);
+                    m_layout.emplace(OutputType::kGloss, 6 * width * height);
+                }
+            }
+        }
+
+        template <class ClType, class Type>
+        void MLDenoiser::ProcessOutput(CLWBuffer<RadeonRays::float3> input,
+                                       Tensor::ValueType* host_mem)
+        {
+            auto input_buf = CLWBuffer<ClType>::CreateFromClBuffer(input);
+            auto normalized_buf = CLWBuffer<ClType>::CreateFromClBuffer(*m_device_cache);
+
+            m_primitives->Normalize(0,
+                                    input_buf,
+                                    normalized_buf,
+                                    (int)input_buf.GetElementCount()).Wait();
+
+            m_context->ReadBuffer<Type>(0,
+                                        CLWBuffer<Type>::CreateFromClBuffer(
+                                        normalized_buf),
+                                        (Type*)(host_mem),
+                                        input_buf.GetElementCount()).Wait();
+        }
+
+        void MLDenoiser::Apply(InputSet const& input_set, Output& output)
+        {
+            auto shape = m_inference->GetInputShape();
+            auto width = std::get<0>(shape);
+            auto height = std::get<1>(shape);
+
+            auto tensor = m_inference->GetInputTensor();
+            auto host_mem = tensor.data();
+
+            for (const auto& input : input_set)
+            {
+                auto type = input.first;
+                auto clw_output = static_cast<ClwOutput*>(input.second);
+                auto device_mem = clw_output->data();
+
+                switch (type)
+                {
+                    case Renderer::OutputType::kColor:
+                    {
+                        auto mem_to_write = host_mem + m_layout[OutputType::kColor];
+                        ProcessOutput<cl_float3, RadeonRays::float3>(device_mem, mem_to_write);
+                        break;
+                    }
+                    case Renderer::OutputType::kDepth:
+                    {
+                        auto mem_to_write = host_mem + m_layout[OutputType::kDepth];
+                        ProcessOutput<cl_float, Tensor::ValueType>(device_mem, mem_to_write);
+                        break;
+                    }
+                    case Renderer::OutputType::kViewShadingNormal:
+                    {
+                        auto mem_to_write = host_mem + m_layout[OutputType::kViewShadingNormal];
```

Comment:
this line can ve place before switch - it;s common for all cases:
`auto mem_to_write = host_mem + m_layout[type];`

Reply:
Done